Mathlib/Topology/Category/Profinite/ExplicitLimits.lean

```diff
@@ -0,0 +1,314 @@
+which are compatible with the maps to the base.
+This is essentially the universal property of the pullback.
+-/
+def pullback.lift {Z : Profinite.{u}} (a : Z ⟶ X) (b : Z ⟶ Y) (w : a ≫ f = b ≫ g) :
+    Z ⟶ pullback f g where
+  toFun := fun z => ⟨⟨a z, b z⟩, by apply_fun (fun q => q z) at w; exact w⟩
+  continuous_toFun := by
+    apply Continuous.subtype_mk
+    rw [continuous_prod_mk]
+    exact ⟨a.continuous, b.continuous⟩
+
+@[reassoc (attr := simp)]
+lemma pullback.lift_fst {Z : Profinite.{u}} (a : Z ⟶ X) (b : Z ⟶ Y) (w : a ≫ f = b ≫ g) :
+  pullback.lift f g a b w ≫ pullback.fst f g = a := rfl
+
+@[reassoc (attr := simp)]
+lemma pullback.lift_snd {Z : Profinite.{u}} (a : Z ⟶ X) (b : Z ⟶ Y) (w : a ≫ f = b ≫ g) :
+  pullback.lift f g a b w ≫ pullback.snd f g = b := rfl
+
+lemma pullback.hom_ext {Z : Profinite.{u}} (a b : Z ⟶ pullback f g)
+    (hfst : a ≫ pullback.fst f g = b ≫ pullback.fst f g)
+    (hsnd : a ≫ pullback.snd f g = b ≫ pullback.snd f g) : a = b := by
+  ext z
+  apply_fun (fun q => q z) at hfst hsnd
+  apply Subtype.ext
+  apply Prod.ext
+  · exact hfst
+  · exact hsnd
+
+/--
+The pullback cone whose cone point is the explicit pullback.
+-/
+@[simps! pt π]
+def pullback.cone : Limits.PullbackCone f g :=
+  Limits.PullbackCone.mk (pullback.fst f g) (pullback.snd f g) (pullback.condition f g)
+
+/--
+The explicit pullback cone is a limit cone.
+-/
+@[simps! lift]
+def pullback.isLimit : Limits.IsLimit (pullback.cone f g) :=
+  Limits.PullbackCone.isLimitAux _
+    (fun s => pullback.lift f g s.fst s.snd s.condition)
+    (fun _ => pullback.lift_fst _ _ _ _ _)
+    (fun _ => pullback.lift_snd _ _ _ _ _)
+    (fun _ _ hm => pullback.hom_ext _ _ _ _ (hm .left) (hm .right))
+
+section Isos
+
+noncomputable
+def ToExplicit : Limits.pullback f g ⟶ Profinite.pullback f g :=
+  pullback.lift f g Limits.pullback.fst Limits.pullback.snd Limits.pullback.condition
+
+noncomputable
+def FromExplicit : Profinite.pullback f g ⟶ Limits.pullback f g :=
+  Limits.pullback.lift (pullback.fst _ _) (pullback.snd _ _) (pullback.condition f g)
+
+@[simp]
+theorem toExplicit_comp_fromExplicit_eq_id :
+    (ToExplicit f g ≫ FromExplicit f g) = 𝟙 _ := by
+  refine' Limits.pullback.hom_ext (k := (ToExplicit f g ≫ FromExplicit f g)) _ _
+  · simp [ToExplicit, FromExplicit]
+  · rw [Category.id_comp, Category.assoc, FromExplicit, Limits.pullback.lift_snd,
+      ToExplicit, pullback.lift_snd]
+
+@[simp]
+theorem fromExplicit_comp_toExplicit_eq_id :
+    (FromExplicit f g ≫ ToExplicit f g) = 𝟙 _ :=
+  pullback.hom_ext f g _ _ (by simp [ToExplicit, FromExplicit]) (by simp [ToExplicit, FromExplicit])
+
+@[simps]
+noncomputable
+def FromExplicitIso : Profinite.pullback f g ≅ Limits.pullback f g where
+  hom := FromExplicit f g
+  inv := ToExplicit f g
+  hom_inv_id := fromExplicit_comp_toExplicit_eq_id f g
+  inv_hom_id := toExplicit_comp_fromExplicit_eq_id f g
+
+end Isos
+
+section Compatibility
+
+theorem explicit_fst_eq :
```

Comment:
I feel it's more natural to state the reversed equality, as it is done below for the coproduct. I don't have a real argument, so if you prefer this order I'm fine with it, but it should be the same for all explicit constructions.

Reply:
I want to have the explicit stuff on the LHS, and the abstract limit stuff on the RHS. This approach is now unified in this PR and the corresponding one for CompHaus